### ISSUE TITLE
Bearriver: More generic definitions

### ISSUE THREAD

**Issue by walseb** [ivanperez-keera/dunai#252]
Many arrows in bearriver don't need to have a notion of time, for example `constant`:
```haskell
constant :: Monad m => b -> SF m a b
constant = arr . const
```
And yet they are defined as a `SF` instead of a `MSF`. Is there any good reason for this? Would PRs generalizing some of these be accepted?

**Comment by walseb** [comment]
After thinking about this some more, one downside to using MSFs in bearriver is that it's more difficult to search for definitions using hoogle. It also blurs the line somewhat between dunai and bearriver. The main upside being the ability to use bearriver primitives in `MSF`s without any notion of time

**Comment by turion** [comment]
Yes, I think that's mainly the reason. A lot of functions like the one you mentioned would belong both in bearriver _and_ dunai though.

**Comment by ivanperez-keera** [comment]
I think it's a fair idea, but I think it's the kind of stuff it's worth thinking about more deeply.

You are probably onto something here.

Every time you generalize the signature, you are also abstracting yourself away from unnecessary details. What you'll likely end up with will be sets of primitives with some additional constraints. So you may have that, if time is ordered, then you can do all of these things, and even without time, you can do such and such.

That's very connected to the idea of the reverse FRP zoo, where we tried to see the commonalities between FRP implementations and express others as extensions and concretizations of MSFs with specific monads.

IMO worth investigating, worth playing around with, and even writing papers about. I would not immediately add these changes to bearriver, but I would not discard having something in bearriver (or between bearriver and dunai) that is more general.

Apart from the concern about hoogle, another issue is the error messages. Although I have not been terribly careful about that, generalizing signatures often makes error messages more abstract and difficult to understand (because they are expressed at a level of abstraction that is different from the DSLs). Food for thought.
